tools/slang-test/slang-test-main.cpp

```diff
@@ -3400,43 +3421,101 @@ static bool shouldRunTest(
     return true;
 }
 
+void getFilesInDirectory(String directoryPath, List<String>& files)
+{
+    {
+        List<String> localFiles;
+        DirectoryUtil::findFiles(directoryPath, localFiles);
+        files.addRange(localFiles);
+    }
+    {
+        List<String> subDirs;
+        DirectoryUtil::findDirectories(directoryPath, subDirs);
+        for (auto subDir : subDirs)
+        {
+            getFilesInDirectory(subDir, files);
+        }
+    }
+}
+
 void runTestsInDirectory(
     TestContext*		context,
     String				directoryPath)
 {
+    List<String> files;
+    getFilesInDirectory(directoryPath, files);
+    auto processFile = [&](String file)
     {
-        List<String> files;
-        DirectoryUtil::findFiles(directoryPath, files);
-        for (auto file : files)
+        if (shouldRunTest(context, file))
         {
-            if( shouldRunTest(context, file) )
+            //            fprintf(stderr, "slang-test: found '%s'\n", file.getBuffer());
+            if (SLANG_FAILED(_runTestsOnFile(context, file)))
             {
-    //            fprintf(stderr, "slang-test: found '%s'\n", file.getBuffer());
-                if (SLANG_FAILED(_runTestsOnFile(context, file)))
                 {
-                    auto reporter = context->reporter;
-
-                    {
-                        TestReporter::TestScope scope(reporter, file);
-                        reporter->message(TestMessageType::RunError, "slang-test: unable to parse test");
+                    TestReporter::TestScope scope(context->getTestReporter(), file);
+                    context->getTestReporter()->message(
+                        TestMessageType::RunError, "slang-test: unable to parse test");
 
-                        reporter->addResult(TestResult::Fail);
-                    }
-
-                    // Output there was some kind of error trying to run the tests on this file
-                    // fprintf(stderr, "slang-test: unable to parse test '%s'\n", file.getBuffer());
+                    context->getTestReporter()->addResult(TestResult::Fail);
                 }
+
+                // Output there was some kind of error trying to run the tests on this file
+                // fprintf(stderr, "slang-test: unable to parse test '%s'\n", file.getBuffer());
             }
         }
+    };
+    bool useMultiThread = false;
+    switch (context->options.defaultSpawnType)
+    {
+    case SpawnType::UseFullyIsolatedTestServer:
+    case SpawnType::UseTestServer:
+        useMultiThread = true;
+        break;
+    }
+    if (context->options.serverCount == 1)
+    {
+        useMultiThread = false;
+    }
+    if (!useMultiThread)
+    {
+        for (auto file : files)
+        {
+            processFile(file);
+        }
```

Comment:
I suspect there is a way we could architect this to avoid some of the tangled complexity that is showing up here.

Ideally the logic in the host/parent process can just be that it walks the directory hierarchy and hands of a sequence of file paths to another component that it *thinks of* as the actual test runner (which it can invoke asynchronously, knowing that status will be reported back eventually). Then we can have a variety of implementations of that test-runner component abstraction:

* A simple test runner that just takes the file paths as they come in and directly tries to execute them as a test (what `processFile` seems to do here?)

* A JSONRPC test runner that maintains a connection with a (single!) sub-process for running tests and sends JSONRPC messages to that process to get it to run the test and report back status.

* A mutli-threaded runner that adds incoming path names to a work queue so that they can be consumed by its pool of worker threads, where each worker thread owns a sub-test-runner that it will pass the path names along to as it grabs them.

I really think we are reaching a point where our test tool has had a *lot* of new functionality bolted onto it, and we haven't always stepped back to make sure that we are keeping the overall architecture of the thing in a reasonable shape.